### ISSUE TITLE
Does kws_streaming support GPU and eager execution?

### ISSUE THREAD

**Issue by ntyoshi** [google-research/google-research#679]
Hi,
I tried to train ds_tc_net by default parameters on GPU.
When I saw GPU process, I could see it used GPU memory but the computation didn't seem to run on GPU.
Do I have to pass some other parameters to the script in order to compute training on GPU?
```
$ echo $CUDA_VISIBLE_DEVICES
3
$ nvidia-smi
Wed Apr 21 10:57:52 2021       
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 460.32.03    Driver Version: 460.32.03    CUDA Version: 11.2     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|                               |                      |               MIG M. |
|===============================+======================+======================|
...
+-------------------------------+----------------------+----------------------+
|   3  Quadro RTX 6000     On   | 00000000:3E:00.0 Off |                  Off |
| 33%   29C    P2    60W / 260W |    167MiB / 24220MiB |      0%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
                                                                               
+-----------------------------------------------------------------------------+
| Processes:                                                                  |
|  GPU   GI   CI        PID   Type   Process name                  GPU Memory |
|        ID   ID                                                   Usage      |
|=============================================================================|
|    3   N/A  N/A     31880      C   python                            163MiB |
+-----------------------------------------------------------------------------+
```

In addition, could you tell me how to try eager execution on kws_streaming project?
I gave it a try to add `run_eagerly = True` to [the line of model.complile](https://github.com/google-research/google-research/blob/34444253e9f57cd03364bc4e50057a5abe9bcf17/kws_streaming/train/train.py#L97) and/or `tf.enable_eager_execution()` before the line but they didn't work.

Sorry for the basic question. Thank you!

**Comment by eggplant95** [comment]
I have the same problems. I found that if I increase the batch size to 1000, only one of the GPUs has increase the Memory usage, yet the GPU_util remains to be 0%. And the training process has been too long. Hope there is a way to increase the GPU usage.

**Comment by rybakov** [comment]
In kws_streaming our focus was mostly on streaming functionality, so you can plug it in any training pipeline. Training loop in this lib was not optimized: there is no data pre-fetching that is why GPU usage will vary in time.
I would suggest to run "watch nvidia-smi" to see how GPU usage is changing in time to validate that it is using GPU.

To increase the gpu usage you will need to modify [data provider](https://source.corp.google.com/piper///depot/google3/third_party/google_research/google_research/kws_streaming/data/input_data.py) and training loop, for example use model.fit() instead of train_on_batch().

We stick with above training loop and data provider for compatibility with [tf example](https://github.com/tensorflow/tensorflow/tree/master/tensorflow/examples/speech_commands) and [hello edge](https://github.com/ARM-software/ML-KWS-for-MCU) work, so that we could compare them side by side.

There are two options for streaming conversion:
1 Based on keras functional api, shown in [models](https://github.com/google-research/google-research/tree/master/kws_streaming/models) folder. It propagates streaming buffer automatically through the model. But it works only in session/graph mode (due to constrains in keras: input layers creation is not allowed inside of a layer in eager mode).

2 Based on keras subclass api, shown in [models_sub](https://github.com/google-research/google-research/tree/master/kws_streaming/models_sub) folder. User will have to manage and propagate streaming buffers manually as show in [function](https://github.com/google-research/google-research/blob/master/kws_streaming/models_sub/conv_model.py#L95). It works in eager mode. But it is not integrated with the training loop of this lib, because current training loop is based on session/graph mode due to [data provider](https://source.corp.google.com/piper///depot/google3/third_party/google_research/google_research/kws_streaming/data/input_data.py) implementation, which can be converted to eager mode too - it is out of scope of our current project.

**Comment by ntyoshi** [comment]
@rybakov 
>I would suggest to run "watch nvidia-smi" to see how GPU usage is changing in time to validate that it is using GPU.
>
>To increase the gpu usage you will need to modify data provider and training loop, for example use model.fit() instead of train_on_batch().

I did `watch nvidia-smi` and confirmed the GPU running a training of kws_streaming kept 0%.
And it seems not to work just by changing from train_on_batch() to fit().
I think it's as you mentioned.

>There are two options for streaming conversion:

I can't see the link of data provider you gave me but do you mean if I want to train some models eagerly with high GPU usage, I need to make a training loop by scratch following 1?
If so, can you give me some example to write?
 (I guess your project has the script that can be an example)

**Comment by rybakov** [comment]
Before optimizing model training with data pre-fetching and modifying the code with model.fit etc, I would suggest to check that your tf and nvidia cuda are set up correctly. Per this [description](https://www.tensorflow.org/install/pip) I installed [tf with gpu support](https://storage.googleapis.com/tensorflow/linux/gpu/tensorflow_gpu-2.4.0-cp36-cp36m-manylinux2010_x86_64.whl) and installed [cuda 11](https://www.tensorflow.org/install/gpu). I run the training and confirm that gpu is used and it is fluctuating between 3% and 50%, and if I increase the batch size (to 512) then peak gpu usage goes to 70%. The changes in gpu usage is due to the lack of data pre-fetching in this lib. When you train the model please check printed logs in the beginning of training: it should say that cuda libs are loaded, if there is any error in *.so loading then training will fall back to cpu training.

**Comment by ttnt-1013** [comment]
@rybakov 
Thank you for your advices. I reviewed my environment especially lib related to CUDA, eventually I could confirm Volatile GPU util working! 
In my case, I use conda as a virtual environment and `conda install tensorflow-gpu` did work for this issue.

And then in terms of eager execution, I'm going to try to convert [models_sub](https://github.com/google-research/google-research/tree/master/kws_streaming/models_sub)  to eager mode. If you have any other info, please kindly share with me. Any information will help me.

**Comment by ntyoshi** [comment]
@rybakov 
Thank you for your advices. I reviewed my environment especially lib related to CUDA, eventually I could confirm Volatile GPU util working!
In my case, I use conda as a virtual environment and conda install tensorflow-gpu did work for this issue.

And then in terms of eager execution, I'm going to try to convert [models_sub](https://github.com/google-research/google-research/tree/master/kws_streaming/models_sub) to eager mode. If you have any other info, please kindly share with me. Any information will help me.

**Comment by rybakov** [comment]
About eager execution, [models_sub](https://github.com/google-research/google-research/tree/master/kws_streaming/models_sub) already eager mode compatible: you could see it in [conv_model_test.py](https://github.com/google-research/google-research/blob/master/kws_streaming/models_sub/conv_model_test.py): eager mode in this file is not disabled. So you can run inference of this model in eager mode.
But training utility is running in session/graph mode because of [input_data](https://github.com/google-research/google-research/blob/master/kws_streaming/data/input_data.py) - it is not eager mode compatible. So if you make [input_data](https://github.com/google-research/google-research/blob/master/kws_streaming/data/input_data.py) eager mode compatible then you could train [models_sub](https://github.com/google-research/google-research/tree/master/kws_streaming/models_sub) in eager mode.

There are also other models in [folder](https://github.com/google-research/google-research/tree/master/kws_streaming/models) but they are all not eager compatible, so eager mode is disabled explicitly in this [example test](https://github.com/google-research/google-research/blob/master/kws_streaming/models/ds_tc_resnet_test.py#L141).

**Comment by ttnt-1013** [comment]
I see.
Do you know any codes I can see as the example of eager mode compatible input?

**Comment by ntyoshi** [comment]
I see.
Do you know any codes I can see as the example of eager mode compatible input?

**Comment by rybakov** [comment]
[input_data](https://github.com/google-research/google-research/blob/master/kws_streaming/data/input_data.py) loads wav files into batch and does some pre-processing. The simplest approach to replace it is to load all training data and validation data and call [model.fit](https://www.tensorflow.org/api_docs/python/tf/keras/Model#fit). This way it will work. Accuracy will not be the same because [input_data](https://github.com/google-research/google-research/blob/master/kws_streaming/data/input_data.py) is pre-processing the data by stretching and shifting audio in time, it also adds noise and adds "unknown" label.

To reproduce the accuracy you could to add learning rate schedule and enable [rand_stretch_squeeze](https://github.com/google-research/google-research/blob/master/kws_streaming/layers/speech_features.py#L233) and [rand_shift](https://github.com/google-research/google-research/blob/master/kws_streaming/layers/speech_features.py#L234) in [speech_features](https://github.com/google-research/google-research/blob/master/kws_streaming/layers/speech_features.py#L234) also will need to add noise to training data and create "unknown" label.

Another option is to convert [input_data](https://github.com/google-research/google-research/blob/master/kws_streaming/data/input_data.py) to eager mode: will need to remove all sessions and validate that it works

**Comment by shikenjo-admin** [comment]
Thank you @rybakov 
Sorry for leaving this issue for a long time.
I understood the situation.
Let me close this issue!

**Comment by ntyoshi** [comment]
Thank you @rybakov
Sorry for leaving this issue for a long time.
I understood the situation.
Let me close this issue!
